### DISCUSSION TITLE
Problema al descargar un fichero

### DISCUSSION THREAD

**Discussion by alexfdzs** [Arquisoft/faq#68] [Q&A]
Hola, soy del equipo wiq_es1b.
Hemos añadido la funcionalidad de descargar todas las preguntas y usuarios de la base de datos en ficheros json.
El problema es el siguiente:
Al llamar a este endpoint (localhost:8000/getAllQuestions o getAllUsers) directamente con la URL funciona perfectamente, pero cuando lo hago desde un botón (que hace axios.get(localhost:8000/getAllQuestions o getAllUsers), se ejecuta el método perfectamente, recopila toda la información y, pese a recibir el mensaje de "Archivo descargado exitosamente", no se descarga nigún archivo.

La rama desde la que se puede probar es [alex's_branch](https://github.com/Arquisoft/wiq_es1b/tree/alex's-branch)

y el código donde se produce el problema está en: 
- [NavigationBar.js](https://github.com/Arquisoft/wiq_es1b/blob/alex's-branch/webapp/src/components/NavigationBar.js#L235-L250) -> Es la vista desde la que se llama al endpoint.
- [Gateway-service.js](https://github.com/Arquisoft/wiq_es1b/blob/alex's-branch/gatewayservice/gateway-service.js#L95-L133) -> Es el controlador que recibe ese endpoint y tiene el código de descarga.
- [Question-service.js](https://github.com/Arquisoft/wiq_es1b/blob/alex's-branch/questionsgenerator/questions-service.js#L53-L63) y [auth-service.js](https://github.com/Arquisoft/wiq_es1b/blob/alex's-branch/users/authservice/auth-service.js#L71-L81) -> Son los servicios que recogen toda la información (el problema no es la información, si no que no se descarga el archivo).

y la manera de reproducir el error es:
1. Arrancar la aplicación con docker (tambien se puede con npm).
2. Añadir un usuario y hacer login con el mismo.
3. Introducir manualmente la URL "localhost:8000/getAllQuestions" o "localhost:8000/getAllUsers". Esta forma es la que funciona.
4. Arriba a la derecha, clicar en el símbolo de info, y seleccionar la opción "get DB data". 
![image](https://github.com/Arquisoft/faq/assets/79599344/48504271-2cc0-4a39-9f4d-2eb71af90276)

5. Se abrirá un diálogo para confirmar la descarga, al clicar en "download", se llamará a ambos endpoints (tanto preguntas como usuarios). En la consola se verá el mensaje de "Archivo descargado exitosamente", pero no se habrá descargado nada.
![image](https://github.com/Arquisoft/faq/assets/79599344/a06b260a-5a3b-4866-a138-118c5e525894)
![image](https://github.com/Arquisoft/faq/assets/79599344/324c0e70-49c7-4d8c-be39-302446cf9f6b)

Si alguien tuviese alguna idea de por qué con la URL sí, pero con axios no, sería de gran ayuda.
Muchas gracias de antemano, un saludo.

WIQ_ES1B

**Comment by pglez82** [comment]
No es tan sencillo como hacer la petición a través de axios sin más. Tenéis que crear un archivo temporal y generar un link para descargarlo (a efectos prácticos el usuario no verá nada):

```javascript
axios({
      url: `${apiEndpoint}/getAllUsers`,
      method: 'GET',
      responseType: 'blob',
    }).then((response) => {
      const href = URL.createObjectURL(response.data);
  
      const link = document.createElement('a');
      link.href = href;
      link.setAttribute('download', 'users.json'); 
      document.body.appendChild(link);
      link.click();
 
      document.body.removeChild(link);
      URL.revokeObjectURL(href);
  });
```
El mismo código sería para el otro archivo también.
